File: ios/Shared/BrokerClient.swift
import Foundation

/// A successful relay fetch together with the broker endpoint that won brokerapi's native race.
public struct BrokerFetch: Sendable {
    public let brokerURL: URL
    public let response: RelayListResponse

    public init(brokerURL: URL, response: RelayListResponse) {
        self.brokerURL = brokerURL
        self.response = response
    }
}

/// Thin Swift projection of brokerapi discovery. Candidate selection, custom-override policy,
/// staggered racing, redirect refusal, signature verification, and ECH transport selection all
/// remain inside the merged Go binding.
public enum BrokerClient {
    public static func firstReachable(
        primary: URL,
        limit: Int = 5,
        clientID: String? = nil,
        sessionID: String? = nil,
        operationFactory: any NativeBrokerOperationFactory
    ) async throws -> BrokerFetch {
        guard let nativeLimit = Int32(exactly: limit) else {
            throw BrokerNativeFailure(
                kind: .validation,
                message: "The relay-list limit is outside the native binding range."
            )
        }

        let result: NativeBrokerRelayResultSnapshot = try await NativeBrokerRunner.run(
            factory: operationFactory
        ) { operation in
            operation.firstReachable(
                primary: primary.absoluteString,
                limit: nativeLimit,
                clientID: clientID ?? "",
                sessionID: sessionID ?? ""
            )
        }
        try result.throwIfFailed()

        guard
            result.brokerURL.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false,
            let winningURL = URL(string: result.brokerURL),
            winningURL.scheme != nil
        else {
            throw BrokerNativeFailure(
                kind: .decode,
                message: "The native broker result contained an invalid winning URL."
            )
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let response: RelayListResponse
        do {
            // Decode the verified bytes exactly once. Decoder errors are deliberately discarded:
            // they can quote attacker-controlled response fields and are not remote diagnostics.
            response = try decoder.decode(
                RelayListResponse.self,
                from: Data(result.relayJSON.utf8)
            )
        } catch {
            throw BrokerNativeFailure(
                kind: .decode,
                message: "The verified relay list is incompatible with this app version."
            )
        }
        return BrokerFetch(brokerURL: winningURL, response: response)
    }
}

File: android/app/src/main/java/com/openrung/net/BrokerClient.kt
package com.openrung.net

import com.openrung.model.RelayListResponse
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json

/** Thin Android adapter over brokerapi's verified, ECH-capable FirstReachable operation. */
object BrokerClient {
    /** A verified relay list together with the exact broker front selected by brokerapi. */
    data class Fetch(val brokerUrl: String, val response: RelayListResponse)

    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Delegates candidate policy, staggered racing, transport selection, and relay verification to
     * brokerapi. Optional identity values become empty strings; Go enforces that a non-empty
     * identity is a complete client/session pair.
     */
    suspend fun firstReachable(
        primary: String,
        limit: Int = 5,
        clientId: String? = null,
        sessionId: String? = null,
    ): Fetch = firstReachable(
        primary = primary,
        limit = limit,
        clientId = clientId,
        sessionId = sessionId,
        operationFactory = NativeBrokerTransport.operationFactory,
    )

    internal suspend fun firstReachable(
        primary: String,
        limit: Int,
        clientId: String?,
        sessionId: String?,
        operationFactory: NativeBrokerOperationFactory,
        decoder: Json = json,
    ): Fetch {
        val native = runNativeBrokerOperation(operationFactory) { operation ->
            operation.firstReachable(
                primary = primary,
                limit = limit,
                clientId = clientId.orEmpty(),
                sessionId = sessionId.orEmpty(),
            )
        }
        val result = requireNativeBrokerSuccess(native, "native broker discovery")
        if (result.brokerUrl.isBlank()) {
            throw BrokerNativeFailure(
                kind = BrokerNativeFailureKind.VALIDATION,
                message = "native broker discovery returned no winning broker URL",
            )
        }
        val response = try {
            decoder.decodeFromString<RelayListResponse>(result.relayJson)
        } catch (_: SerializationException) {
            // Parser diagnostics can quote the relay payload. Keep the local incompatibility fixed,
            // typed, and free of response bytes; Go has already selected and verified this winner.
            throw BrokerNativeFailure(
                kind = BrokerNativeFailureKind.DECODE,
                message = "native broker relay JSON is incompatible with the Android model",
            )
        }
        return Fetch(brokerUrl = result.brokerUrl, response = response)
    }
}
